kubernetes-sigs/sig-storage-local-static-provisioner
# Update discovery and deletion code to work in Windows nodes through CSI Proxy

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind feature

**What this PR does / why we need it**:
Runs the discover and cleanup process in Windows nodes through CSI Proxy, implementation details:

- A disk must be partitioned and volumes must be mounted to the same location as the linux component (in the `C:/` drive) 

## Implementation details

- The config `hostDir` and `volumeDir` are escaped to be under `C:/` when the config is read, this is a requirement for the user to do i.e. **the user must mount their volumes to the C:/ drive in the same location as in linux**
- There are a few implementation of volumeUtils for CSI Proxy v1beta and v1, I might remove the v1beta version later but during testing I used v1beta
- Volume capacity is computed with CSI Proxy (`GetVolumeIDFromTargetPath` -> `GetVolumeStats`)
- Volume deletion is implemented with CSI Proxy (`GetVolumeIDFromTargetPath` -> `FormatVolume`)

## Testing

- Create a GKE cluster with 1 linux node and 1 windows node, each node will have 2 local SSDs
- In the windows node:
  - Initialize and disk and create a partition, mount the partition to the same location as the linux node (e.g. `C:\mnt\disks\ssd0` and `C:\mnt\disks\ssd1`)
- Build and deploy LVP in the cluster, we'll have 2 pods corresponding to the daemonsets for Linux and Windows, I deployed the `helm/generated-examples/development-gke.yaml` file
- Verify that LVP can connect to CSI Proxy (v1beta preinstalled in the GKE node)
- Verify that 2 PV are created for the Windows nodes, verify that their reported capacity is correct (corresponds to CSI Proxy's `GetVolumeStats` results)
- Create a test workload with a PVC that uses the same StorageClass as the LVP setup
- Verify that one of the PVs is bound to the PVC and to the workload
- Write something in the workload's mounted filesystem, verify that the files are in the host's filesystem too
- Delete the workload and the PVC
- Verify that the deleter reformats the mounted volume before deleting the PV

**Which issue(s) this PR fixes**:

Fixes #271
Fixes #269


**Release note**:
```
Update discovery and deletion code to work in Windows nodes through CSI Proxy, after a PV gets cleaned by local volume provisioner it'll reformat the backing volume with NTFS 
```

/cc @jingxu97
